File: src/main/java/com/sothawo/akkabatch/java/RecordModifier.java
package com.sothawo.akkabatch.java;

import akka.actor.ActorSelection;
import com.sothawo.akkabatch.scala.messages.ProcessRecord;
import com.typesafe.config.ConfigException;

import java.text.MessageFormat;
import java.util.Random;

/**
 * RecordModifier Actor.
 *
 * @author P.J. Meisch (pj.meisch@sothawo.com).
 */
public class RecordModifier extends AkkaBatchActor {
// ------------------------------ FIELDS ------------------------------

    protected static final String CONFIG_DROPRATE = "simulation.recordModifier.droprate";

    private static Random random = new Random();
    /** the Writer */
    private ActorSelection writer;
    /** drop rate per thousand */
    private int dropRatePerMille = 0;
    private long numProcessed;
    private long numDropped;

// ------------------------ CANONICAL METHODS ------------------------

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof ProcessRecord) {
            boolean drop = false;
            if (dropRatePerMille > 0 && random.nextInt(1000) < dropRatePerMille) {
                drop = true;
            }
            if (!drop) {
                ProcessRecord in = (ProcessRecord) message;
                ProcessRecord out = new ProcessRecord(in.getRecordId(), in.getCsvOriginal(),
                                                      RecordProcessor.processRecord(in.getRecord()));
                writer.tell(out, getSelf());
                numProcessed++;
            } else {
                numDropped++;
            }
        } else {
            unhandled(message);
        }
    }

    @Override
    public void preStart() throws Exception {
        super.preStart();
        try {
            dropRatePerMille = appConfig.getInt(CONFIG_DROPRATE);
        } catch (ConfigException e) {
            log.error(e, CONFIG_DROPRATE);
        }

        numProcessed = 0;
        numDropped = 0;

        // Writer ist im Master
        String writerPath = appConfig.getString("network.master.address") + appConfig.getString("names.writerRef");
        log.info("Writer path from configuration: " + writerPath);
        writer = context().actorSelection(writerPath);
        log.info(MessageFormat.format("sending data to {0}, drop rate: {1} 0/00", writer.pathString(), dropRatePerMille));
    }

    @Override
    public void postStop() throws Exception {
        log.debug(MessageFormat.format("processed: {0}, dropped: {1}", numProcessed, numDropped));
        super.postStop();
    }
}

File: src/main/scala/com/sothawo/akkabatch/scala/RecordModifier.scala
package com.sothawo.akkabatch.scala

import akka.actor.ActorSelection
import com.typesafe.config.ConfigException
import com.sothawo.akkabatch.scala.messages.ProcessRecord
import scala.util.Random

/**
 * @author P.J. Meisch (pj.meisch@sothawo.com).
 */
class RecordModifier extends AkkaBatchActor {

  protected final val CONFIG_DROPRATE: String = "simulation.recordModifier.droprate"

  /** the Writer */
  private var writer: ActorSelection = null

  /** drop rate per thousand */
  private var dropRatePerMille: Int = 0

  private var numProcessed: Long = _

  private var numDropped: Long = _

  override def preStart(): Unit = {
    super.preStart()

    try {
      dropRatePerMille = appConfig.getInt(CONFIG_DROPRATE)
    } catch {
      case e: ConfigException => log.error(e, CONFIG_DROPRATE)
    }

    numProcessed = 0
    numDropped = 0

    // Writer is in Master
    val writerPath: String = appConfig.getString("network.master.address") + appConfig.getString("names.writerRef")
    writer = context.actorSelection(writerPath)

    log.info(s"Writer path from configuration: $writerPath")
    log.info(s"sending data to ${writer.pathString}, drop rate: $dropRatePerMille 0/00")
  }

  override def postStop() {
    log.debug(s"processed: $numProcessed, dropped: $numDropped")
    super.postStop()
  }

  def receive = {
    case msg: ProcessRecord => {
      if (dropRatePerMille == 0 || Random.nextInt(1000) >= dropRatePerMille) {
        writer ! ProcessRecord(msg.recordId, msg.csvOriginal, RecordProcessor.processRecord(msg.record))
        numProcessed += 1
      } else {
        numDropped += 1
      }
    }
  }

}
